src/promise.c

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include "crt.h"
+
+#include <aws/common/condition_variable.h>
+#include <aws/common/mutex.h>
+
+struct _aws_crt_promise {
+    struct aws_mutex mutex;
+    struct aws_condition_variable cv;
+    bool complete;
+    int error_code;
+    void *value;
+    /* destructor for value, will be invoked if the value is not taken */
+    void (*dtor)(void *);
+};
+
+aws_crt_promise *aws_crt_promise_new(void) {
+    aws_crt_promise *promise = aws_crt_mem_acquire(sizeof(aws_crt_promise));
+    memset(promise, 0, sizeof(*promise));
+    aws_mutex_init(&promise->mutex);
+    aws_condition_variable_init(&promise->cv);
+    return promise;
+}
+
+void aws_crt_promise_delete(aws_crt_promise *promise) {
```

Comment:
Gotta admit I love the destructors that check null so that all the callers can just unconditionally call it.  Really nice for RAII cleanup in constructors.